### PR TITLE
update pidcontroller for 2019

### PR DESCRIPTION
**Pull request by ariovistus**
depends on #490

Fixes: #38

### CODE REVIEW

File: wpilib/wpilib/pidbase.py [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+from .resource import Resource
+from .timer import Timer
+from .lineardigitalfilter import LinearDigitalFilter
+
+__all__ = ["PIDBase"]
+
+class PIDBase(SendableBase):
+    """Class implements a PID Control Loop.
+
+    Creates a separate thread which reads the given PIDSource and takes care of the integral
```
Comment (auscompgeek) [inline wpilib/wpilib/pidbase.py]: Heh, this seems wrong; this class doesn't spawn a thread. But it's what upstream has. I guess we should file a bug, I suppose.

Comment (ariovistus) [reply wpilib/wpilib/pidbase.py]: my impression was it leaves the choice of notifier vs non notifier up to the subclass

Comment (auscompgeek) [reply wpilib/wpilib/pidbase.py]: I think that's the intention here, yes. But that doesn't change the fact that there's a documentation bug here.

**Comment by auscompgeek** [comment]
Rebase this and we'll merge, I guess.

File: wpilib/wpilib/pidbase.py [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+        """
+        Wraps error around for continuous inputs. The original error is
+        returned if continuous mode is disabled. This is an unsynchronized
+        function.
+
+        :param error: The current error of the PID controller.
+        :return: Error for continuous inputs.
+        """
+        if self.continuous and self.inputRange > 0:
+            error %= self.inputRange
```
Comment (auscompgeek) [inline wpilib/wpilib/pidbase.py]: You know, I keep staring at this line. I know you didn't originally write this, but the meaning of `%` in Python is different to other C-like languages, so `error` is always positive here. The rough equivalent of C/Java's `%` is `math.fmod()`.

I sat down at 1am thinking this out, and I'm pretty sure the following is equivalent to both what upstream is doing and what this is doing:

```python
error %= self.inputRange
if error > self.inputRange / 2:
    return error - self.inputRange
```

Comment (ariovistus) [reply wpilib/wpilib/pidbase.py]: Pretty sure I looked at that and wondered about it, but figured "meh, it's in the existing codebase, it's probably fine"

Comment (auscompgeek) [reply wpilib/wpilib/pidbase.py]: Yeah. I'm surprised we haven't ported across the PID tolerance tests.

File: wpilib/wpilib/pidbase.py [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+    # is on target.  The various implementations of this such as
+    # PercentageTolerance and AbsoluteTolerance specify types of tolerance
+    # specifications to use.
+    def PercentageTolerance_onTarget(self, percentage):
+        return abs(self.getError()) < percentage / 100.0 * self.inputRange
+
+    def AbsoluteTolerance_onTarget(self, value):
+        return abs(self.getError()) < value
+
+    def __init__(self, Kp, Ki, Kd, Kf, output, source):
```
Comment (auscompgeek) [inline wpilib/wpilib/pidbase.py]: Java has a default for Kf here. Do you think it's worth putting it here too? We can make source and output keyword-only arguments.

Wait, source and output are in the wrong order compared to Java too.

Comment (ariovistus) [reply wpilib/wpilib/pidbase.py]: I think it's not worth putting on the base

Comment (auscompgeek) [reply wpilib/wpilib/pidbase.py]: The argument order still needs to be fixed though.

Having the default here could simplify the PIDController constructor too.

**Comment by auscompgeek** [comment]
Needs another rebase since we merged the type hints PR. Please ensure the type hints are preserved in PIDBase as well.

**Pull request merged**